src/authorization/PermissionBasedAuthorizer.ts

```diff
@@ -16,7 +16,8 @@ import { AccessMode } from './permissions/Permissions';
  * if yes authorization is granted.
  * `undefined` values for reader results are interpreted as `false`.
  */
-export class PermissionBasedAuthorizer extends Authorizer {
```

Comment:
But what is the special relevance of the three original fields then?

Reply:
Just to help with typing when these fields are used, and give indicators of what can be used. Depending on what is needed it might make more sense to also extend those sub objects similarly.

I'm also OK with adding whatever field is needed for UMA to this object btw, as UMA is part of the specification.

Reply:
I actually think that the 4 generic type changes I made are all we need as abstraction for UMA. Do you still think using Records is overkill? If not, I'd prefer to keep it that way for modularity; then I can play around with the exact implementation of UMA credentials, rather than fixing their type in the core code.

Reply:
Fixing the UMA thing in the core code was just a suggestion if it could be useful. I don't know enough about UMA to know if there are fixed values that are always relevant, or if it can differ between tokens. If it is the latter it does indeed not make much sense to have it in the core.

My issue with the Records as it is written now is that it makes the `Credentials` type less relevant. Now the code is "take an object with credential info, of which `Credentials` is the default choice", while I would prefer it to be "take an object that extends `Credentials`". The idea of the `Credentials` type when it was made, was that it was the baseline of which credentials could be provided, but that it could also be extended if necessary.

I'm not sure what you need, but this is something that is already possible:
```ts
interface MyCredentials extends Credentials {
  uma: { a: number; b: string };
};
```
which would be a valid object to pass to the `Authorizer`, which does not need to be aware of what exactly is in the `Credentials` object anyway. So does this not already allow for adding anything you want to your own custom credentials object and passing it along to the necessary classes?

Reply:
> The idea of the Credentials type when it was made, was that it was the baseline of which credentials could be provided

I understand, but what I'm trying to point out is that such a baseline is not useful, since different authentication approaches use completely different (non-overlapping) fields. Even more, since the introduction of UMA means that Solid servers will no longer have to handle OIDC themselves, the original 'baseline' fields will become completely irrelevant with time.

Reply:
> [An extension of Credentials] would be a valid object to pass to the Authorizer, which does not need to be aware of what exactly is in the Credentials object anyway.

Not entirely true, I think, since in the AuthorizingHttpHandler, we will need some way to couple the Credential output type of the CredentialsExtractor with the Credential input type of the PermissionReader. Or am I missing something?

Reply:
> Not entirely true, I think, since in the `AuthorizingHttpHandler`, we will need some way to couple the Credential output type of the `CredentialsExtractor` with the `Credential` input type of the `PermissionReader`. Or am I missing something?

That depends on how you do type guards. The `AuhtorizingHttpHandler` doesn't necessarily need to know the exact type of what it passes to the `PermissionReader`. It is just there to pass values along. Due to how Components.js is used to instantiate everything, generic type guards like that don't have much value anyway as they don't actually get checked during instantiation. If you want to have type guards, the solution is generally to do this in the class that cares about the more specific type. E.g., the UMA class would see if the `Credentials` that it receives contains the necessary fields, and if this is not the case it would throw an error. Depending on how the full architecture is, potentially another class would then be called that can handle this type of credentials, and if there isn't one the error would be propagated to the output.

> I understand, but what I'm trying to point out is that such a baseline is not useful

My point is that if this is the case, there is an issue with the `Credentials` type and potentially it has to change. It is the type that is returned by a `CredentialsExtractor` and ingested by a `PermissionReader` (and an `Authorizer` to know if we throw a 401/403, but less relevant there), so instead of changing those classes to accept more than `Credentials`, I would prefer to change `Credentials` so it matches everything needed by those classes. We would then still need a solution to make sure the classes extracting the WebID put this result somewhere so the other classes know where to find it.

This entire architecture, of the `AuthorizingHttpHandler` doing the steps of extracting credentials/modes/permissions and comparing the outputs, was designed with the "standard" WAC/ACP authorization methods in mind, where doing those specific steps, and expecting `Credentials` to contain things like a WebID, makes sense.

Do those steps still make sense for UMA as well? Because an alternative solution could be to have a different HTTP handler that replaces the `AuthorizingHttpHandler` that is UMA specific should the requirements differ too much. This would depend on what exactly these changes are needed for and what the plans are for UMA support. Is this only for the step of the RS validating the token? Or is this also the AS part of generating a token?

This does also feel like it's becoming a bit much for a discussion on a PR comment 😅.

Also note that some solutions might be breaking, so a solution could be to have a temporary solution here that is good enough together with an issue of fully changing this in the next major.